### ISSUE TITLE
deIndex is being called wrong

### ISSUE THREAD

**Issue by wraithgar** [AmpersandJS/ampersand-collection#45]
https://github.com/AmpersandJS/ampersand-collection/blob/master/ampersand-collection.js#L310-L311

Pretty sure we want to send the old values to `_deIndex`

```javascript
    _onModelEvent: function (event, model, emitter, options) {
        if ((event === 'add' || event === 'remove') && emitter !== this) return;
        if (event === 'destroy') this.remove(model, options);
        if (model && event === 'change:' + this.mainIndex) {
            this._deIndex(emitter);
            this._index(model);
        }
        this.trigger.apply(this, arguments);
    }
```

**Comment by wraithgar** [comment]
Also, what happens if any other attribute in this.indexes change?

**Comment by wraithgar** [comment]
Maybe
```javascript
    _onModelEvent: function (event, model, emitter, options) {
        var attribute = event.split(':')[1];
        event = event.split(':')[0];
        if ((event === 'add' || event === 'remove') && emitter !== this) return;
        if (event === 'destroy') this.remove(model, options);
        if (model && event === 'change' && this._indexes[attribute]) {
            this._deIndex(emitter);
            this._index(model);
        }
       
```

**Comment by chabou** [comment]
In case of single attribute change, `emitter` is the new value of this attribute.
Maybe we could use previousAttributes()
```
var attribute = event.split(':')[1];
        event = event.split(':')[0];
        if ((event === 'add' || event === 'remove') && emitter !== this) return;
        if (event === 'destroy') this.remove(model, options);
        if (model && event === 'change' && this._indexes[attribute]) {
            this._deIndex(model.previousAttributes());
            this._index(model);
        }
```

**Comment by wraithgar** [comment]
`this._deIndex(model, attribute, value)`

This would require deIndex to be given the index value too, much like it can be given the specific index/

**Comment by chabou** [comment]
ok but maybe optional parameters. If they aren't passed to the method, we loop on all model attributes. (useful for _addReference)

We could do the same for _index

**Comment by chabou** [comment]
Can I fill a PR ? Even if you don't accept my changes, I added some useful test cases.

**Comment by wraithgar** [comment]
@chabou this issue is 100% intended for someone who has the time to write up a PR for this THANK YOU SO MUCH if you have the time.

**Comment by lukekarrys** [comment]
Closed by #47
